### PR TITLE
io_uring: add io_uring socket handle

### PR DESCRIPTION
**Pull request by zhxie**
Signed-off-by: He Jie Xu <hejie.xu@intel.com>
Signed-off-by: Xie Zhihao <zhihao.xie@intel.com>

Commit Message: io_uring: add io_uring socket handle
Additional Description: This is a pretty large (mostly in tests) patch which introduces `IoUringSocketHandle`. `IoUringSocketHandle` benefits from `IoUringWorker`, `IoUringWorkerFactory` and `IoUringSocket` and exposes io_uring features to the upper components within the `IoHandle` API. `IoUringSocketHandle` does only support UDP-related features. In the next patch, we will define the API and merge io_uring into the socket interface to make io_uring work in Envoy.
Risk Level: Low
Testing: Unit and integration tests
Docs Changes: N/A (will be added when the socket interface and its API are ready) 
Release Notes: N/A
Platform Specific Features: N/A
Part of #28395
Fix #32265

### CODE REVIEW

**Comment by zhxie** [comment]
@phlax We have added some guards to avoid including io_uring in Android builds. So is there a way to enable the Android CI again?

**Comment by soulxu** [comment]
> @phlax We have added some guards to avoid including io_uring in Android builds. So is there a way to enable the Android CI again?

It seems like here https://github.com/envoyproxy/envoy/blob/9e447bfaa2eaf29925bfb279b83e1ad202d1b514/.github/config.yml#L168-L176

Maybe you can add one line there in this PR and give a test.

I guess we needn't test the iouring with Envoy mobile(not only android). So maybe just debug in this PR is enough.

**Comment by zhxie** [comment]
> > @phlax We have added some guards to avoid including io_uring in Android builds. So is there a way to enable the Android CI again?
> 
> It seems like here (I wasn't very sure)
> 
> https://github.com/envoyproxy/envoy/blob/9e447bfaa2eaf29925bfb279b83e1ad202d1b514/.github/config.yml#L168-L176
> 
> Maybe you can add one line there in this PR and give a test.
> 
> I guess we needn't test the iouring with Envoy mobile(not only android). So maybe just debug in this PR is enough.

Good point. I prefer to change something in these paths to trigger the CI test and then revert the change back.

**Comment by zhxie** [comment]
Checks against [74aa3b7](https://github.com/envoyproxy/envoy/pull/32946/commits/74aa3b78370e86a79b36fdea48d049c235c37380) shows there is no issue around mobile build now.
@soulxu @yanavlasov @KBaichoo can you re-review the PR? The only change from the original one is https://github.com/envoyproxy/envoy/pull/32946/commits/98e2955e3bfa0658693c5562837eaa1e6728c7c4.
Also CC @phlax

File: source/common/network/BUILD
Comment (soulxu) [inline source/common/network/BUILD:256]: Does this mean we still build those iouring code for mobile? I'm curious that why mobile coverage test is ok for this.

Comment (zhxie) [reply source/common/network/BUILD:256]: It look like `//bazel:linux` only builds in the real Linux (else it may fail the CI).

**Comment by phlax** [comment]
> I prefer to change something in these paths to trigger the CI test and then revert the change back.

@zhxie yep you need to do this - changing `config.yml` wont work in a PR

File: source/common/network/BUILD
Comment (zhxie) [reply source/common/network/BUILD:256]: Yea, the CI also passes after removing the file.

**Review by soulxu — DISMISSED** [review]
so it seems the envoy mobile won't compile the stuff of iouring, so LGTM, thanks!

**Comment by zhxie** [comment]
> > I prefer to change something in these paths to trigger the CI test and then revert the change back.
> 
> @zhxie yep you need to do this - changing `config.yml` wont work in a PR

Yes, it has been tested, and the CI passed.

File: source/common/network/io_uring_socket_handle_impl.cc
Comment (RyanTheOptimist) [inline source/common/network/io_uring_socket_handle_impl.cc:32]: nit: remove a layer of nesting by doing:
``` 
if (!SOCKET_VALID(fd_)) {
  return;
}
...
```

File: source/common/network/io_uring_socket_handle_impl.cc
Comment (RyanTheOptimist) [inline source/common/network/io_uring_socket_handle_impl.cc:45]: nit: Is it "better" to use close, or is it required?

File: source/common/network/io_uring_socket_handle_impl.cc
Comment (RyanTheOptimist) [inline source/common/network/io_uring_socket_handle_impl.cc:89]: nit: please use the fully type name here, per the style guide:

The fundamental rule is: use type deduction only to make the code clearer or safer, and do not use it merely to avoid the inconvenience of writing an explicit type. 
https://google.github.io/styleguide/cppguide.html#Type_deduction

File: source/common/network/io_uring_socket_handle_impl.cc
Comment (RyanTheOptimist) [inline source/common/network/io_uring_socket_handle_impl.cc:105]: ditto about the full type. (here and elsewhere)

File: source/common/network/io_uring_socket_handle_impl.cc
Comment (RyanTheOptimist) [inline source/common/network/io_uring_socket_handle_impl.cc:183]: Full type please.

**Review by RyanTheOptimist — COMMENTED** [review]
I'm trusting @soulxu for domain knowledge here so mostly doing a general review. LGTM other than a few nits.

**Comment by alyssawilk** [comment]
hm, looks like the mobile checks are still skipped despite the workflow change?  cc @phlax and can you just touch something in Envoy Mobile (I often punctuate a test comment) to actually run the tests?

**Comment by phlax** [comment]
> hm, looks like the mobile checks are still skipped despite the workflow change?

if you check commits above i think it was run previously (by touching `mobile/README`)

File: source/common/network/socket_interface_impl.cc
Comment (zhxie) [reply source/common/network/socket_interface_impl.cc]: Fixed, thanks.

File: source/common/network/io_uring_socket_handle_impl.cc
Comment (zhxie) [reply source/common/network/io_uring_socket_handle_impl.cc]: Fixed, thanks.

File: source/common/network/io_uring_socket_handle_impl.cc
Comment (zhxie) [reply source/common/network/io_uring_socket_handle_impl.cc]: Required. rm `better`.

File: source/common/network/io_uring_socket_handle_impl.cc
Comment (zhxie) [reply source/common/network/io_uring_socket_handle_impl.cc]: Well, if you want the full type, then I can just give you the full type. But I think the result from `accept` is also `auto` even in the core io_socket_handle_impl.cc.

**Comment by alyssawilk** [comment]
Let's leave a mobile file touch until the dot file change is synced and active - I would definitely like this PR to keep running mobile CI until it lands

**Comment by RyanTheOptimist** [comment]
LGTM, but let's make sure we do what @alyssawilk is asking for before final approval.

**Comment by zhxie** [comment]
Trigger mobile CI. Let's wait for checks.

**Review by RyanTheOptimist — APPROVED** [review]

**Comment by RyanTheOptimist** [comment]
LGTM. Shall I merge it?

**Comment by zhxie** [comment]
> LGTM. Shall I merge it?
> 
> 

I will remove the test-only commit after the final approval, then we can begin to merge. I am not sure if there is any further review.

**Comment by RyanTheOptimist** [comment]
AH, right. I think @soulxu and I have both approve, so once you remove the test-only commit we're good to go.

**Comment by zhxie** [comment]
Done. @soulxu @RyanTheOptimist can you give the approval again so we can start to merge.

**Review by RyanTheOptimist — APPROVED** [review]

**Review by soulxu — APPROVED** [review]
LGTM, thanks

**Pull request merged**
